### PR TITLE
Feature/ruler

### PR DESCRIPTION
**Pull request by owen-d**
## What

Introduces the Cortex Ruler to Loki. 

This enables the often-requested feature of alerting based on log entries. It's Prometheus & [cortextool](https://github.com/grafana/cortex-tools#cortextool) compatible. Ruling allows alerting via expressions such as 
image

## Details

This PR is based largely upon a linked PR in cortex which enables it. We build atop those changes, introducing an in memory series store in order to hold past evaluations of a rule. This allows Loki to be run independently from cortex/prometheus and keep track of how long an alert has been active. 

I've vendored my fork for clarity while the dependent PR is open in cortex.

## Depends on

https://github.com/cortexproject/cortex/pull/2812

## Further work for later PRs

- Pluggable metric stores to remove the need to re-evaluate past rule evaluations in order to populate data for the `For` clause after restarts or ring re-sharding. The most obvious candidate here is Cortex.
- Allow configurable granularity around evaluating past rules after a restart. If we try to prepoulate the past evaluations at the default evaluation interval of 1 minute for everyone  rule afer restarting, we expose ourselves to overburdening the read path.

### CODE REVIEW

File: pkg/logql/ast.go [hunk 1]
```diff
@@ -21,6 +21,11 @@ type Expr interface {
 	fmt.Stringer
 }
 
+// implicit holds default implementations
```
Comment (owen-d) [inline pkg/logql/ast.go:24]: this is a totally unrelated change that I snuck in.

**Comment by AllenzhLi** [comment]
Why not abstract a common ruler which can be sharing by both cortex and loki？
